File: webapp/src/lib/render/video-engine.ts
/**
 * Video Engine for generating lyrics videos.
 *
 * Generates MP4 videos with synchronized lyrics using frame rendering
 * and FFmpeg encoding. Supports multiple templates and resolutions.
 */

import * as fs from "fs/promises";
import { existsSync } from "fs";
import * as path from "path";
import ffmpeg from "fluent-ffmpeg";
import ffmpegStatic from "ffmpeg-static";
import { spawn } from "child_process";
import { AssetFetcher } from "./asset-fetcher";
import {
  FrameRenderer,
  VideoTemplate,
  VideoTemplateName,
  FontSizePreset,
  SegmentInfo,
} from "./frame-renderer";
import {
  parseLRC,
  GlobalLRCLine,
  convertToGlobalTimeline,
} from "./lrc-parser";
import { chaptersToFFmpegMetadata, ChaptersManifest } from "./chapters";
import { AudioSegmentInfo } from "./audio-engine";

export interface VideoEngineOptions {
  template?: VideoTemplateName;
  fontSizePreset?: FontSizePreset;
  resolution?: "720p" | "1080p";
  fps?: number;
  includeTitleCard?: boolean;
  titleCardDurationSeconds?: number;
  ffmpegPath?: string;
  ffprobePath?: string;
}

export interface VideoExportResult {
  outputPath: string;
  totalFrames: number;
  durationSeconds: number;
  width: number;
  height: number;
  fps: number;
}

export interface ChapterInfo {
  position: number;
  songTitle: string;
  startSeconds: number;
  endSeconds: number;
  lines: Array<{
    text: string;
    startSeconds: number;
  }>;
}

export type ProgressCallback = (currentFrame: number, totalFrames: number) => void;

/**
 * VideoEngine generates lyrics videos synchronized with audio.
 */
export class VideoEngine {
  private assetFetcher: AssetFetcher;
  private template: VideoTemplate;
  private fontSizePreset: FontSizePreset;
  private resolution: { width: number; height: number };
  private fps: number;
  private includeTitleCard: boolean;
  private titleCardDurationSeconds: number;
  private ffmpegPath: string;
  private ffprobePath: string;
  private frameRenderer: FrameRenderer;

  constructor(
    assetFetcher: AssetFetcher,
    options: VideoEngineOptions = {}
  ) {
    this.assetFetcher = assetFetcher;
    this.template = FrameRenderer.getTemplate(options.template ?? "dark");
    this.fontSizePreset = options.fontSizePreset ?? "M";
    this.resolution =
      options.resolution === "720p"
        ? { width: 1280, height: 720 }
        : { width: 1920, height: 1080 };
    this.fps = options.fps ?? 24;
    this.includeTitleCard = options.includeTitleCard ?? true;
    this.titleCardDurationSeconds = Math.min(
      Math.max(options.titleCardDurationSeconds ?? 5, 5),
      30
    );
    this.ffmpegPath = options.ffmpegPath ?? (ffmpegStatic && existsSync(ffmpegStatic) ? ffmpegStatic : "ffmpeg");
    this.ffprobePath = options.ffprobePath ?? "ffprobe";

    this.frameRenderer = new FrameRenderer({
      template: this.template,
      fontSizePreset: this.fontSizePreset,
      resolution: this.resolution,
    });
  }

  /**
   * Initialize the video engine.
   */
  async initialize(): Promise<void> {
    await this.frameRenderer.initialize();
  }

  /**
   * Generate a lyrics video synchronized with audio.
   *
   * @param audioPath - Path to the mixed audio file
   * @param segments - Audio segment information with song metadata
   * @param outputPath - Path for output video
   * @param progressCallback - Called with (currentFrame, totalFrames)
   * @returns VideoExportResult with output information
   */
  async generateVideo(
    audioPath: string,
    segments: AudioSegmentInfo[],
    outputPath: string,
    progressCallback?: ProgressCallback
  ): Promise<VideoExportResult> {
    // Ensure output directory exists
    await fs.mkdir(path.dirname(outputPath), { recursive: true });

    // Get audio duration
    const audioInfo = await this.getAudioInfo(audioPath);
    if (!audioInfo) {
      throw new Error("Could not get audio info");
    }

    const totalDurationSeconds = audioInfo.durationSeconds;
    const titleCardFrames = this.includeTitleCard
      ? Math.ceil(this.titleCardDurationSeconds * this.fps)
      : 0;
    const totalFrames = Math.ceil(totalDurationSeconds * this.fps) + titleCardFrames;

    // Collect all lyrics with global timing
    const allLyrics: GlobalLRCLine[] = [];
    const chapters: ChapterInfo[] = [];

    for (let i = 0; i < segments.length; i++) {
      const segment = segments[i];
      const hashPrefix = segment.item.recordingHashPrefix;

      if (!hashPrefix) {
        continue;
      }

      // Download and parse LRC
      const lrcContent = await this.assetFetcher.downloadLrc(hashPrefix);
      if (!lrcContent) {
        console.warn(`No LRC found for ${hashPrefix}`);
        continue;
      }

      const localLyrics = parseLRC(lrcContent);
      const globalLyrics = convertToGlobalTimeline(
        localLyrics,
        segment.startTimeSeconds,
        segment.item.songTitle ?? segment.item.songId?.toString() ?? `song-${i}`
      );

      allLyrics.push(...globalLyrics);

      // Build chapter info
      const segmentEnd = segment.startTimeSeconds + segment.durationSeconds;
      chapters.push({
        position: i + 1,
        songTitle: segment.item.songTitle ?? segment.item.songId?.toString() ?? `Song ${i + 1}`,
        startSeconds: segment.startTimeSeconds,
        endSeconds: segmentEnd,
        lines: localLyrics.map((line) => ({
          text: line.text,
          startSeconds: segment.startTimeSeconds + line.timeSeconds,
        })),
      });
    }

    if (allLyrics.length === 0) {
      // Generate blank video if no lyrics
      return this.generateBlankVideo(audioPath, outputPath, totalDurationSeconds);
    }

    // Convert segments to SegmentInfo format
    const segmentInfos: SegmentInfo[] = segments.map((seg, i) => ({
      id: seg.item.id,
      songId: seg.item.songId,
      position: seg.item.position,
      songTitle: seg.item.songTitle ?? seg.item.songId?.toString() ?? `Song ${i + 1}`,
      startTimeSeconds: seg.startTimeSeconds,
      durationSeconds: seg.durationSeconds,
      tempoBpm: seg.item.tempoBpm,
    }));

    const titleCardConfig = this.includeTitleCard
      ? {
          frames: titleCardFrames,
          songsetName: segments[0]?.item.songTitle ?? "Worship Set",
          songCount: segments.length,
          totalDurationSeconds,
        }
      : undefined;

    // Generate video with FFmpeg
    await this.encodeVideoWithFFmpeg(
      audioPath,
      outputPath,
      totalFrames,
      totalDurationSeconds,
      allLyrics,
      segmentInfos,
      progressCallback,
      titleCardConfig
    );

    return {
      outputPath,
      totalFrames,
      durationSeconds: totalDurationSeconds,
      width: this.resolution.width,
      height: this.resolution.height,
      fps: this.fps,
    };
  }

  /**
   * Encode video using FFmpeg with raw video input.
   */
  private async encodeVideoWithFFmpeg(
    audioPath: string,
    outputPath: string,
    totalFrames: number,
    totalDurationSeconds: number,
    lyrics: GlobalLRCLine[],
    segments: SegmentInfo[],
    progressCallback?: ProgressCallback,
    titleCardConfig?: {
      frames: number;
      songsetName: string;
      songCount: number;
      totalDurationSeconds: number;
    }
  ): Promise<void> {
    const { width, height } = this.resolution;

    // Build FFmpeg command
    const args = [
      "-y", // Overwrite output
      "-f",
      "rawvideo",
      "-vcodec",
      "rawvideo",
      "-s",
      `${width}x${height}`,
      "-pix_fmt",
      "rgba",
      "-r",
      String(this.fps),
      "-i",
      "-", // Read from stdin
      "-i",
      audioPath,
      ...this.getVideoCodecArgs(),
      "-c:a",
      "aac",
      "-b:a",
      "192k",
      "-shortest",
      outputPath,
    ];

    return new Promise((resolve, reject) => {
      const process = spawn(this.ffmpegPath, args, {
        stdio: ["pipe", "ignore", "ignore"],
      });

      let frameCount = 0;
      let isProcessing = true;

      const writeFrame = async () => {
        if (!isProcessing) return;

        try {
          let buffer: Buffer;

          if (titleCardConfig && frameCount < titleCardConfig.frames) {
            const canvas = this.frameRenderer.renderTitleCard({
              enabled: true,
              durationSeconds: titleCardConfig.totalDurationSeconds,
              songsetName: titleCardConfig.songsetName,
              songCount: titleCardConfig.songCount,
              totalDurationSeconds: titleCardConfig.totalDurationSeconds,
            });
            buffer = this.frameRenderer.canvasToBuffer(canvas);
          } else {
            const lyricsFrameIndex = titleCardConfig ? frameCount - titleCardConfig.frames : frameCount;
            const currentTime = lyricsFrameIndex / this.fps;

            const canvas = this.frameRenderer.renderFrame(
              lyrics,
              segments,
              currentTime
            );
            buffer = this.frameRenderer.canvasToBuffer(canvas);
          }

          if (process.stdin && process.stdin.writable) {
            const canContinue = process.stdin.write(buffer, (err) => {
              if (err) {
                isProcessing = false;
                reject(err);
                return;
              }

              frameCount++;

              if (progressCallback && frameCount % this.fps === 0) {
                progressCallback(frameCount, totalFrames);
              }

              if (frameCount >= totalFrames) {
                if (process.stdin) {
                  process.stdin.end();
                }
              }
            });

            if (!canContinue && isProcessing && frameCount < totalFrames) {
              await new Promise<void>((resolve) => {
                process.stdin!.once("drain", resolve);
              });
            }

            if (frameCount < totalFrames && isProcessing && canContinue) {
              setImmediate(writeFrame);
            } else if (frameCount < totalFrames && isProcessing) {
              setImmediate(writeFrame);
            }
          }
        } catch (error) {
          isProcessing = false;
          reject(error);
        }
      };

      process.on("error", (err) => {
        isProcessing = false;
        reject(err);
      });

      process.on("exit", (code) => {
        isProcessing = false;
        if (code === 0) {
          if (progressCallback) {
            progressCallback(totalFrames, totalFrames);
          }
          resolve();
        } else {
          reject(new Error(`FFmpeg exited with code ${code}`));
        }
      });

      // Start writing frames
      writeFrame();
    });
  }

  /**
   * Generate a blank video with just the background color.
   */
  private async generateBlankVideo(
    audioPath: string,
    outputPath: string,
    durationSeconds: number
  ): Promise<VideoExportResult> {
    const { width, height } = this.resolution;
    const [bgR, bgG, bgB] = this.template.backgroundColor;

    const args = [
      "-y",
      "-f",
      "lavfi",
      "-i",
      `color=c=#${bgR.toString(16).padStart(2, "0")}${bgG.toString(16).padStart(2, "0")}${bgB.toString(16).padStart(2, "0")}:s=${width}x${height}:d=${durationSeconds}`,
      "-i",
      audioPath,
      ...this.getVideoCodecArgs("5000k"),
      "-c:a",
      "aac",
      "-b:a",
      "192k",
      "-shortest",
      outputPath,
    ];

    await new Promise<void>((resolve, reject) => {
      const process = spawn(this.ffmpegPath, args);

      process.on("error", reject);
      process.on("exit", (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`FFmpeg exited with code ${code}`));
        }
      });
    });

    return {
      outputPath,
      totalFrames: Math.ceil(durationSeconds * this.fps),
      durationSeconds,
      width,
      height,
      fps: this.fps,
    };
  }

  /**
   * Get platform-appropriate video codec arguments.
   */
  private getVideoCodecArgs(bitrate: string = "8000k"): string[] {
    // Use libx264 with ultrafast preset for best compatibility
    return [
      "-c:v",
      "libx264",
      "-preset",
      "ultrafast",
      "-crf",
      "23",
      "-b:v",
      bitrate,
    ];
  }

  /**
   * Get audio file information using ffprobe.
   */
  private async getAudioInfo(
    filePath: string
  ): Promise<{
    durationSeconds: number;
    durationMs: number;
    sampleRate: number;
    channels: number;
  } | null> {
    return new Promise((resolve, reject) => {
      ffmpeg.ffprobe(filePath, (err, metadata) => {
        if (err) {
          reject(err);
          return;
        }

        if (!metadata.streams || metadata.streams.length === 0) {
          reject(new Error("No audio streams found"));
          return;
        }

        const stream = metadata.streams[0];
        const durationSeconds = metadata.format.duration ?? 0;

        resolve({
          durationSeconds,
          durationMs: Math.round(durationSeconds * 1000),
          sampleRate: stream.sample_rate ?? 44100,
          channels: stream.channels ?? 2,
        });
      });
    });
  }

  /**
   * Inject chapter atoms into MP4 file.
   * Best-effort: proceeds on failure.
   *
   * @param videoPath - Path to MP4 file
   * @param chapters - Chapter information
   * @returns True if successful
   */
  async injectChapters(
    videoPath: string,
    chapters: ChapterInfo[]
  ): Promise<boolean> {
    try {
      // Create temporary file for chapters
      const tempDir = await this.assetFetcher.getTempDir();
      const chaptersPath = path.join(tempDir, `chapters-${Date.now()}.txt`);

      // Write chapters in FFmpeg metadata format
      const manifest: ChaptersManifest = {
        chapters,
        totalDurationSeconds: 0,
        generatedAt: new Date().toISOString(),
      };
      const chaptersContent = chaptersToFFmpegMetadata(manifest);
      await fs.writeFile(chaptersPath, chaptersContent, "utf-8");

      // Create output path
      const outputPath = `${videoPath}.chapters.mp4`;

      // Run FFmpeg to inject chapters
      const args = [
        "-y",
        "-i",
        videoPath,
        "-i",
        chaptersPath,
        "-map_metadata",
        "1",
        "-c",
        "copy",
        outputPath,
      ];

      await new Promise<void>((resolve, reject) => {
        const process = spawn(this.ffmpegPath, args);
        process.on("error", reject);
        process.on("exit", (code) => {
          if (code === 0) {
            resolve();
          } else {
            reject(new Error(`FFmpeg exited with code ${code}`));
          }
        });
      });

      // Replace original file with chapter-injected version
      await fs.rename(outputPath, videoPath);

      // Clean up temp file
      await fs.unlink(chaptersPath).catch(() => {});

      return true;
    } catch (error) {
      console.warn("Failed to inject chapters (proceeding anyway):", error);
      return false;
    }
  }

}

File: services/render-worker/src/sow_render_worker/video_engine.py
from __future__ import annotations

import logging
import math
import shutil
import subprocess
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Protocol

from sow_render_worker.audio_engine import AudioSegmentInfo, get_audio_info
from sow_render_worker.chapters import Chapter, ChaptersManifest, chapters_to_ffmpeg_metadata
from sow_render_worker.frame_renderer import (
    VIDEO_TEMPLATES,
    FontSizePreset,
    FrameRenderer,
    SegmentInfo,
    TitleCardConfig,
    VideoTemplateName,
)
from sow_render_worker.lrc_parser import GlobalLRCLine, convert_to_global_timeline, parse_lrc

logger = logging.getLogger(__name__)


RESOLUTION_MAP: dict[str, tuple[int, int]] = {
    "720p": (1280, 720),
    "1080p": (1920, 1080),
}


@dataclass(frozen=True)
class VideoExportResult:
    output_path: str
    total_frames: int
    duration_seconds: float
    width: int
    height: int
    fps: int


@dataclass(frozen=True)
class ChapterInfo:
    position: int
    song_title: str
    start_seconds: float
    end_seconds: float
    lines: tuple[dict[str, Any], ...] = field(default_factory=tuple)


ProgressCallback = Callable[[int, int], None]
TimeoutCheckCallback = Callable[[], None]


class AssetFetcherProtocol(Protocol):
    def download_lrc(self, hash_prefix: str) -> str | None: ...

    def get_temp_dir(self) -> Path: ...


class VideoEngine:
    def __init__(
        self,
        asset_fetcher: AssetFetcherProtocol,
        template: VideoTemplateName = "dark",
        font_size_preset: FontSizePreset = "M",
        resolution: str = "1080p",
        fps: int = 24,
        include_title_card: bool = True,
        title_card_duration_seconds: float = 5.0,
        ffmpeg_path: str | None = None,
        ffprobe_path: str | None = None,
    ):
        self.asset_fetcher = asset_fetcher
        self.template = VIDEO_TEMPLATES.get(template, VIDEO_TEMPLATES["dark"])
        self.font_size_preset = font_size_preset
        self.resolution = RESOLUTION_MAP.get(resolution, (1920, 1080))
        self.fps = fps
        self.include_title_card = include_title_card
        self.title_card_duration_seconds = max(5.0, min(title_card_duration_seconds, 30.0))
        self.ffmpeg_path = ffmpeg_path or self._find_ffmpeg()
        self.ffprobe_path = ffprobe_path or "ffprobe"

        self.frame_renderer = FrameRenderer(
            template=self.template,
            font_size_preset=self.font_size_preset,
            resolution=self.resolution,
        )

    @staticmethod
    def _find_ffmpeg() -> str:
        found = shutil.which("ffmpeg")
        return found or "ffmpeg"

    def get_video_codec_args(self, bitrate: str = "8000k") -> list[str]:
        return [
            "-c:v",
            "libx264",
            "-preset",
            "ultrafast",
            "-crf",
            "23",
            "-b:v",
            bitrate,
        ]

    def generate_video(
        self,
        audio_path: str,
        segments: list[AudioSegmentInfo],
        output_path: str,
        progress_callback: ProgressCallback | None = None,
        timeout_check_callback: TimeoutCheckCallback | None = None,
        job_id: str | None = None,
    ) -> VideoExportResult:
        output_dir = Path(output_path).parent
        output_dir.mkdir(parents=True, exist_ok=True)

        audio_info = get_audio_info(audio_path)
        if not audio_info:
            raise ValueError("Could not get audio info")

        total_duration_seconds = audio_info["duration_seconds"]
        title_card_frames = (
            math.ceil(self.title_card_duration_seconds * self.fps)
            if self.include_title_card
            else 0
        )
        total_frames = math.ceil(total_duration_seconds * self.fps) + title_card_frames

        logger.info(
            "[%s] generate_video: duration=%.1fs, total_frames=%d, resolution=%s, fps=%d",
            job_id or "unknown", total_duration_seconds, total_frames,
            f"{self.resolution[0]}x{self.resolution[1]}", self.fps,
        )

        all_lyrics: list[GlobalLRCLine] = []
        chapters: list[ChapterInfo] = []

        for i, segment in enumerate(segments):
            hash_prefix = segment.item.recording_hash_prefix
            if not hash_prefix:
                continue

            try:
                lrc_content = self.asset_fetcher.download_lrc(hash_prefix)
            except Exception:
                lrc_content = None
            if not lrc_content:
                continue

            local_lyrics = parse_lrc(lrc_content)
            title = (
                segment.item.song_title
                or (str(segment.item.song_id) if segment.item.song_id else f"song-{i}")
            )
            global_lyrics = convert_to_global_timeline(
                local_lyrics, segment.start_time_seconds, title
            )
            all_lyrics.extend(global_lyrics)

            segment_end = segment.start_time_seconds + segment.duration_seconds
            chapters.append(
                ChapterInfo(
                    position=i + 1,
                    song_title=(
                        segment.item.song_title
                        or (str(segment.item.song_id) if segment.item.song_id else f"Song {i + 1}")
                    ),
                    start_seconds=segment.start_time_seconds,
                    end_seconds=segment_end,
                    lines=tuple(
                        {
                            "text": line.text,
                            "startSeconds": segment.start_time_seconds + line.time_seconds,
                        }
                        for line in local_lyrics
                    ),
                )
            )

        if not all_lyrics:
            return self.generate_blank_video(
                audio_path, output_path, total_duration_seconds,
                job_id=job_id,
            )

        segment_infos: list[SegmentInfo] = []
        for i, seg in enumerate(segments):
            segment_infos.append(
                SegmentInfo(
                    id=seg.item.id,
                    song_id=seg.item.song_id,
                    position=seg.item.position,
                    song_title=(
                        seg.item.song_title
                        or (str(seg.item.song_id) if seg.item.song_id else f"Song {i + 1}")
                    ),
                    start_time_seconds=seg.start_time_seconds,
                    duration_seconds=seg.duration_seconds,
                    tempo_bpm=seg.item.tempo_bpm,
                )
            )

        title_card_config: TitleCardConfig | None = None
        if self.include_title_card:
            title_card_config = TitleCardConfig(
                enabled=True,
                duration_seconds=total_duration_seconds,
                songset_name=(
                    (segments[0].item.song_title or "Worship Set") if segments else "Worship Set"
                ),
                song_count=len(segments),
                total_duration_seconds=total_duration_seconds,
            )

        self.encode_video_with_ffmpeg(
            audio_path,
            output_path,
            total_frames,
            total_duration_seconds,
            all_lyrics,
            segment_infos,
            progress_callback,
            title_card_config,
            timeout_check_callback,
            job_id=job_id,
        )

        logger.info(
            "[%s] generate_video: complete, %d frames encoded",
            job_id or "unknown", total_frames,
        )

        return VideoExportResult(
            output_path=output_path,
            total_frames=total_frames,
            duration_seconds=total_duration_seconds,
            width=self.resolution[0],
            height=self.resolution[1],
            fps=self.fps,
        )

    def encode_video_with_ffmpeg(
        self,
        audio_path: str,
        output_path: str,
        total_frames: int,
        total_duration_seconds: float,
        lyrics: list[GlobalLRCLine],
        segments: list[SegmentInfo],
        progress_callback: ProgressCallback | None = None,
        title_card_config: TitleCardConfig | None = None,
        timeout_check_callback: TimeoutCheckCallback | None = None,
        job_id: str | None = None,
    ) -> None:
        width, height = self.resolution

        ffmpeg_start = time.monotonic()
        logger.info(
            "[%s] encode_video_with_ffmpeg: starting FFmpeg pipe, %d frames (%.1fs at %dfps)",
            job_id or "unknown", total_frames, total_duration_seconds, self.fps,
        )

        args = [
            self.ffmpeg_path,
            "-y",
            "-f",
            "rawvideo",
            "-vcodec",
            "rawvideo",
            "-s",
            f"{width}x{height}",
            "-pix_fmt",
            "rgba",
            "-r",
            str(self.fps),
            "-i",
            "-",
            "-i",
            audio_path,
            *self.get_video_codec_args(),
            "-c:a",
            "aac",
            "-b:a",
            "192k",
            "-shortest",
            output_path,
        ]

        process = subprocess.Popen(
            args,
            stdin=subprocess.PIPE,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.PIPE,
        )

        stderr_chunks: list[bytes] = []
        stderr_thread: threading.Thread | None = None
        if process.stderr:
            def _drain_stderr(pipe, chunks):
                try:
                    while True:
                        chunk = pipe.read(4096)
                        if not chunk:
                            break
                        chunks.append(chunk)
                except Exception:
                    pass

            stderr_thread = threading.Thread(target=_drain_stderr, args=(process.stderr, stderr_chunks), daemon=True)
            stderr_thread.start()

        title_card_frame_count = (
            math.ceil(self.title_card_duration_seconds * self.fps)
            if title_card_config and title_card_config.enabled
            else 0
        )

        title_card_bytes: bytes | None = None
        if title_card_config and title_card_frame_count > 0:
            title_card_img = self.frame_renderer.render_title_card(title_card_config)
            title_card_bytes = title_card_img.tobytes()

        frame_count = 0

        try:
            while frame_count < total_frames:
                if timeout_check_callback:
                    timeout_check_callback()

                if title_card_config and frame_count < title_card_frame_count:
                    frame_bytes = title_card_bytes
                else:
                    lyrics_frame_index = (
                        frame_count - title_card_frame_count if title_card_config else frame_count
                    )
                    current_time = lyrics_frame_index / self.fps
                    img = self.frame_renderer.render_frame(lyrics, segments, current_time)
                    frame_bytes = img.tobytes()

                try:
                    process.stdin.write(frame_bytes)
                except BrokenPipeError:
                    logger.error(
                        "[%s] FFmpeg pipe broken at frame %d/%d",
                        job_id or "unknown", frame_count, total_frames,
                    )
                    process.stdin.close()
                    if stderr_thread:
                        stderr_thread.join(timeout=5)
                    process.wait()
                    stderr_output = b"".join(stderr_chunks).decode("utf-8", errors="replace")
                    stderr_info = (
                        f"\nFFmpeg stderr (last 2000 chars): {stderr_output[-2000:]}"
                        if stderr_output
                        else ""
                    )
                    raise RuntimeError(
                        f"FFmpeg process closed prematurely (EPIPE on stdin write).{stderr_info}"
                    )

                frame_count += 1

                if progress_callback and frame_count % self.fps == 0:
                    progress_callback(frame_count, total_frames)

                if frame_count % (self.fps * 30) == 0 and frame_count > 0:
                    video_seconds = frame_count / self.fps
                    logger.info(
                        "[%s] Video encoding progress: %.0fs/%.0fs (%d/%d frames, %.1f%%)",
                        job_id or "unknown", video_seconds, total_duration_seconds,
                        frame_count, total_frames,
                        frame_count / total_frames * 100 if total_frames > 0 else 0,
                    )

            try:
                process.stdin.close()
            except BrokenPipeError:
                pass
        except Exception:
            process.kill()
            if stderr_thread:
                stderr_thread.join(timeout=5)
            process.wait()
            raise

        if stderr_thread:
            stderr_thread.join(timeout=5)
        return_code = process.wait()
        ffmpeg_elapsed = time.monotonic() - ffmpeg_start
        logger.info(
            "[%s] FFmpeg process exited with code %d in %.1fs",
            job_id or "unknown", return_code, ffmpeg_elapsed,
        )
        stderr_output = b"".join(stderr_chunks).decode("utf-8", errors="replace")
        if return_code != 0:
            stderr_info = (
                f"\nFFmpeg stderr (last 2000 chars): {stderr_output[-2000:]}"
                if stderr_output
                else ""
            )
            raise RuntimeError(f"FFmpeg exited with code {return_code}.{stderr_info}")

        if progress_callback:
            progress_callback(total_frames, total_frames)

    def generate_blank_video(
        self,
        audio_path: str,
        output_path: str,
        duration_seconds: float,
        job_id: str | None = None,
    ) -> VideoExportResult:
        width, height = self.resolution
        bg_r, bg_g, bg_b = self.template.background_color
        hex_color = f"#{bg_r:02x}{bg_g:02x}{bg_b:02x}"

        logger.info(
            "[%s] generate_blank_video: %.1fs, %s",
            job_id or "unknown", duration_seconds, output_path,
        )

        args = [
            self.ffmpeg_path,
            "-y",
            "-f",
            "lavfi",
            "-i",
            f"color=c={hex_color}:s={width}x{height}:d={duration_seconds}",
            "-i",
            audio_path,
            *self.get_video_codec_args("5000k"),
            "-c:a",
            "aac",
            "-b:a",
            "192k",
            "-shortest",
            output_path,
        ]

        result = subprocess.run(
            args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.PIPE,
        )
        if result.returncode != 0:
            stderr_output = result.stderr.decode("utf-8", errors="replace")
            stderr_info = (
                f"\nFFmpeg stderr (last 2000 chars): {stderr_output[-2000:]}"
                if stderr_output
                else ""
            )
            raise RuntimeError(f"FFmpeg exited with code {result.returncode}.{stderr_info}")

        return VideoExportResult(
            output_path=output_path,
            total_frames=math.ceil(duration_seconds * self.fps),
            duration_seconds=duration_seconds,
            width=width,
            height=height,
            fps=self.fps,
        )

    def inject_chapters(
        self,
        video_path: str,
        chapters: list[ChapterInfo],
        job_id: str | None = None,
    ) -> bool:
        logger.info(
            "[%s] inject_chapters: %d chapters into %s",
            job_id or "unknown", len(chapters), video_path,
        )
        try:
            temp_dir = self.asset_fetcher.get_temp_dir()
            chapters_path = str(Path(temp_dir) / f"chapters-{int(time.time() * 1000)}.txt")

            manifest = ChaptersManifest(
                chapters=tuple(
                    Chapter(
                        position=ch.position,
                        song_title=ch.song_title,
                        start_seconds=ch.start_seconds,
                        end_seconds=ch.end_seconds,
                    )
                    for ch in chapters
                ),
                total_duration_seconds=0,
                generated_at="",
            )
            chapters_content = chapters_to_ffmpeg_metadata(manifest)
            Path(chapters_path).write_text(chapters_content, encoding="utf-8")

            output_path = f"{video_path}.chapters.mp4"

            args = [
                self.ffmpeg_path,
                "-y",
                "-i",
                video_path,
                "-i",
                chapters_path,
                "-map_metadata",
                "1",
                "-c",
                "copy",
                output_path,
            ]

            result = subprocess.run(
                args,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.PIPE,
            )

            if result.returncode != 0:
                logger.warning("[%s] inject_chapters: failed", job_id or "unknown")
                return False

            shutil.move(output_path, video_path)

            try:
                Path(chapters_path).unlink()
            except OSError:
                pass

            logger.info("[%s] inject_chapters: complete", job_id or "unknown")
            return True
        except Exception:
            logger.warning("[%s] inject_chapters: failed", job_id or "unknown")
            return False
